File: jscomp/test/test_seq.js
'use strict';

var Curry = require("../../lib/js/curry.js");
var Caml_obj = require("../../lib/js/caml_obj.js");
var Pervasives = require("../../lib/js/pervasives.js");
var Caml_exceptions = require("../../lib/js/caml_exceptions.js");
var Caml_js_exceptions = require("../../lib/js/caml_js_exceptions.js");

var Bad = Caml_exceptions.create("Test_seq.Bad");

var Help = Caml_exceptions.create("Test_seq.Help");

var Stop = Caml_exceptions.create("Test_seq.Stop");

function assoc3(x, _l) {
  while(true) {
    var l = _l;
    if (l) {
      var match = l._0;
      if (Caml_obj.caml_equal(match[0], x)) {
        return match[1];
      }
      _l = l._1;
      continue ;
    }
    throw {
          RE_EXN_ID: "Not_found",
          Error: new Error()
        };
  };
}

function help_action(param) {
  throw {
        RE_EXN_ID: Stop,
        _1: {
          tag: /* Unknown */0,
          _0: "-help"
        },
        Error: new Error()
      };
}

function v(speclist) {
  assoc3("-help", speclist);
  return /* [] */0;
}

function f(g, speclist) {
  return Curry._1(g, assoc3("-help", speclist));
}

function add_help(speclist) {
  var add1;
  try {
    assoc3("-help", speclist);
    add1 = /* [] */0;
  }
  catch (raw_exn){
    var exn = Caml_js_exceptions.internalToOCamlException(raw_exn);
    if (exn.RE_EXN_ID === "Not_found") {
      add1 = /* :: */{
        _0: [
          "-help",
          {
            tag: /* Unit */0,
            _0: help_action
          },
          " Display this list of options"
        ],
        _1: /* [] */0
      };
    } else {
      throw exn;
    }
  }
  var add2;
  try {
    assoc3("--help", speclist);
    add2 = /* [] */0;
  }
  catch (raw_exn$1){
    var exn$1 = Caml_js_exceptions.internalToOCamlException(raw_exn$1);
    if (exn$1.RE_EXN_ID === "Not_found") {
      add2 = /* :: */{
        _0: [
          "--help",
          {
            tag: /* Unit */0,
            _0: help_action
          },
          " Display this list of options"
        ],
        _1: /* [] */0
      };
    } else {
      throw exn$1;
    }
  }
  return Pervasives.$at(speclist, Pervasives.$at(add1, add2));
}

exports.Bad = Bad;
exports.Help = Help;
exports.Stop = Stop;
exports.assoc3 = assoc3;
exports.help_action = help_action;
exports.v = v;
exports.f = f;
exports.add_help = add_help;
/* No side effect */

File: jscomp/test/test_seq.ml
(***********************************************************************)
(*                                                                     *)
(*                                OCaml                                *)
(*                                                                     *)
(*             Damien Doligez, projet Para, INRIA Rocquencourt         *)
(*                                                                     *)
(*  Copyright 1996 Institut National de Recherche en Informatique et   *)
(*  en Automatique.  All rights reserved.  This file is distributed    *)
(*  under the terms of the GNU Library General Public License, with    *)
(*  the special exception on linking described in file ../LICENSE.     *)
(*                                                                     *)
(***********************************************************************)

type key = string
type doc = string
type usage_msg = string
type anon_fun = string -> unit

type spec =
  | Unit of (unit -> unit) (* Call the function with unit argument *)
  | Bool of (bool -> unit) (* Call the function with a bool argument *)
  | Set of bool ref (* Set the reference to true *)
  | Clear of bool ref (* Set the reference to false *)
  | String of (string -> unit) (* Call the function with a string argument *)
  | Set_string of string ref (* Set the reference to the string argument *)
  | Int of (int -> unit) (* Call the function with an int argument *)
  | Set_int of int ref (* Set the reference to the int argument *)
  | Float of (float -> unit) (* Call the function with a float argument *)
  | Set_float of float ref (* Set the reference to the float argument *)
  | Tuple of spec list (* Take several arguments according to the spec list *)
  | Symbol of string list * (string -> unit)
  (* Take one of the symbols as argument and call the function with the symbol. *)
  | Rest of (string -> unit)

(* Stop interpreting keywords and call the function with each remaining
   argument *)

exception Bad of string
exception Help of string

type error =
  | Unknown of string
  | Wrong of string * string * string (* option, actual, expected *)
  | Missing of string
  | Message of string

exception Stop of error

(* used internally *)

open Printf

let rec assoc3 x l =
  match l with
  | [] -> raise Not_found
  | (y1, y2, y3) :: t when y1 = x -> y2
  | _ :: t -> assoc3 x t

(* let make_symlist prefix sep suffix l = *)
(* match l with *)
(* | [] -> "<none>" *)
(* | h::t -> (List.fold_left (fun x y -> x ^ sep ^ y) (prefix ^ h) t) ^ suffix *)
(* ;; *)

(* let print_spec buf (key, spec, doc) = *)
(* if String.length doc > 0 then *)
(* match spec with *)
(* | Symbol (l, _) -> *)
(* bprintf buf " %s %s%s\n" key (make_symlist "{" "|" "}" l) doc *)
(* | _ -> *)
(* bprintf buf " %s %s\n" key doc *)
(* ;; *)

let help_action () = raise (Stop (Unknown "-help"))

let v speclist =
  ignore (assoc3 "-help" speclist) ;
  []

let f g speclist = g (assoc3 "-help" speclist)

let add_help speclist =
  let add1 =
    try
      ignore (assoc3 "-help" speclist) ;
      []
    with Not_found ->
      [("-help", Unit help_action, " Display this list of options")]
  and add2 =
    try
      ignore (assoc3 "--help" speclist) ;
      []
    with Not_found ->
      [("--help", Unit help_action, " Display this list of options")] in
  speclist @ add1 @ add2

(** FIXME- not compatible with strict mode *)

(* let f x x = x *)
